### PR TITLE
 Refactor trainer and introduce stage 2 trainer

### PR DESCRIPTION
**Pull request by alecokas**
Basically addresses issue #36

### CODE REVIEW

File: trainers/stage_2_trainer.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+import numpy as np
+from tqdm import trange
+import tensorflow as tf
+from trainers.base_trainer import Trainer
+from utils.data_helpers import tensors_from_sample
+
+
+class Stage1Trainer(Trainer):
```
Comment (Devin-Taylor) [inline trainers/stage_2_trainer.py:8]: Stage2Trainer?

File: trainers/stage_2_trainer.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                    sample, batch_size, text_embedding_size, self.num_samples, self.augment
+                )
+
+                with tf.GradientTape() as generator_tape, tf.GradientTape() as discriminator_tape:
+                    # Forward pass the stage 1 generator to obtain small fake images
+                    noise_z = tf.random.normal((batch_size, self.noise_size))
+                    fake_images_small, _, _ = self.stage_1_generator([text_tensor, noise_z], training=False)
+
+                    fake_images_large, mean, log_sigma = self.model.generator(
+                        [text_tensor, noise_z, fake_images_small], training=True
```
Comment (Devin-Taylor) [inline trainers/stage_2_trainer.py:52]: yeah just checked, no noise. Only need to pass in [fake_images_small, text_tensor] in that order.

File: trainers/stage_2_trainer.py (hunk 1 above)
Comment (alecokas) [reply trainers/stage_2_trainer.py:8]: Yup lol

File: trainers/stage_2_trainer.py (hunk 2 above)
Comment (alecokas) [reply trainers/stage_2_trainer.py:52]: Perfect, thanks. Their figure 2 is quite misleading then

Comment (Devin-Taylor) [reply trainers/stage_2_trainer.py:52]: stage2
the only things I see going into stage 2 are the embedding and generated images unless I am missing something?

Comment (alecokas) [reply trainers/stage_2_trainer.py]: Inside the conditioning augmentation block there is a noise generation mechanism if we use the same definition as given in stage 1

Comment (Devin-Taylor) [reply trainers/stage_2_trainer.py]: oh ok yeah, but that is not noise_z though

File: models/discriminators.py [hunk 3]
```diff
@@ -125,63 +125,52 @@ def __init__(self, img_size, kernel_size, num_filters, lr, w_init, bn_init):
     def build(self, input_size):
         activation = lambda l: tf.nn.leaky_relu(l, alpha=0.2)
 
-        self.conv_1 = Conv2D(filters=self.d_dim, kernel_size=(4, 4), strides=(2, 2), padding='same',
-                             kernel_initializer=self.w_init, use_bias=False
-                             )
+        self.conv_block_1 = ConvBlock(filters=self.d_dim, kernel_size=(4, 4), strides=(2, 2), padding='same',
+                                      w_init=self.w_init, bn_init=self.bn_init, activation=activation)
```
Comment (Devin-Taylor) [inline models/discriminators.py:129]: There is no batchnorm on the first conv which I why I just did it separately. Should probably change it back

File: models/discriminators.py [hunk 4]
```diff
@@ -125,63 +125,52 @@ def __init__(self, img_size, kernel_size, num_filters, lr, w_init, bn_init):
...
+        self.conv_block_9 = ConvBlock(filters=self.d_dim*8, kernel_size=(1, 1), strides=(1, 1), padding='same',
+                                      w_init=self.w_init, bn_init=self.bn_init, activation=activation)
 
-        # (16, 16) == 256/16
-        self.conv_2 = Conv2D(filters=1, kernel_size=(16, 16), strides=(16, 16),
-                             padding='same', kernel_initializer=self.w_init
-                             )
+        # (4, 4) == 256/16
+        self.conv_block_10 = ConvBlock(filters=1, kernel_size=(4, 4), strides=(4, 4), padding='same',
+                                       w_init=self.w_init, bn_init=self.bn_init, activation=activation)
```
Comment (Devin-Taylor) [inline models/discriminators.py:161]: same here, no batchnorm, activation, and bias needs to be enabled

**Comment by alecokas** [comment]
Changes made and ready for review again

File: models/discriminators.py [hunk 5]
```diff
@@ -125,8 +125,8 @@ def __init__(self, img_size, kernel_size, num_filters, lr, w_init, bn_init):
     def build(self, input_size):
         activation = lambda l: tf.nn.leaky_relu(l, alpha=0.2)
 
-        self.conv_block_1 = ConvBlock(filters=self.d_dim, kernel_size=(4, 4), strides=(2, 2), padding='same',
-                                      w_init=self.w_init, bn_init=self.bn_init, activation=activation)
+        self.conv_1 = Conv2D(filters=self.d_dim, kernel_size=(4, 4), strides=(2, 2), padding='same',
+                             kernel_initializer=self.w_init, use_bias=False, activation=activation)
```
Comment (Devin-Taylor) [inline models/discriminators.py:129]: does passing a tf.nn activation function work?

File: models/discriminators.py [hunk 6]
```diff
@@ -157,13 +157,13 @@ def build(self, input_size):
                                       w_init=self.w_init, bn_init=self.bn_init, activation=activation)
 
         # (4, 4) == 256/16
-        self.conv_block_10 = ConvBlock(filters=1, kernel_size=(4, 4), strides=(4, 4), padding='same',
-                                       w_init=self.w_init, bn_init=self.bn_init, activation=activation)
+        self.conv_2 = Conv2D(filters=1, kernel_size=(4, 4), strides=(4, 4), padding='same',
+                             kernel_initializer=self.w_init, use_bias=False)
```
Comment (Devin-Taylor) [inline models/discriminators.py:161]: should disable bias in stage 1 too

File: models/discriminators.py (hunk 5 above)
Comment (alecokas) [reply models/discriminators.py:129]: Yeah good point. No error gets thrown but I just checked the docs and it doesn't say anything about the tf.nn activations. I guess I'm not sure then. Safer to pass in the standard tf.keras one

File: models/discriminators.py (hunk 6 above)
Comment (alecokas) [reply models/discriminators.py:161]: I believe it is already done

File: models/discriminators.py (hunk 5 above)
Comment (Devin-Taylor) [reply models/discriminators.py:129]: na the keras one throws warnings when we save. Just apply it separately in the forward pass (call function)

File: models/discriminators.py (hunk 6 above)
Comment (alecokas) [reply models/discriminators.py:161]: I believe that it is
